modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -184,60 +224,116 @@ private CompletableFuture<?> onIndexRemoved(RemoveIndexEventParameters parameter
+                    CatalogTableDescriptor tableDescriptor = catalog.table(primaryReplicaId.tableId());
+                    if (tableDescriptor == null) {
+                        return nullCompletedFuture();
+                    }
+
+                    return getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId.tableId())
+                            .thenCompose(mvTableStorage -> awaitPrimaryReplica(primaryReplicaId, parameters.startTime())
+                                    .thenAccept(replicaMeta -> tryScheduleBuildIndexesForNewPrimaryReplica(
+                                            catalog.version(),
+                                            tableDescriptor,
+                                            primaryReplicaId,
+                                            mvTableStorage,
+                                            replicaMeta
+                                    ))
+                            );
+                }
             } else {
-                stopBuildingIndexesIfPrimaryExpired(primaryReplicaId);
+                stopBuildingIndexesIfPrimaryExpired(parameters.groupId());
 
                 return nullCompletedFuture();
             }
         });
     }
 
     private void tryScheduleBuildIndexesForNewPrimaryReplica(
+            // TODO It seems that we can pass the catalog itself instead of its version.
             int catalogVersion,
-            TablePartitionId primaryReplicaId,
+            CatalogTableDescriptor tableDescriptor,
+            ReplicationGroupId primaryReplicaId,
             MvTableStorage mvTableStorage,
             ReplicaMeta replicaMeta
     ) {
         inBusyLock(busyLock, () -> {
-            if (isLeaseExpire(replicaMeta)) {
+            if (isLeaseExpired(replicaMeta)) {
                 stopBuildingIndexesIfPrimaryExpired(primaryReplicaId);
 
                 return;
             }
 
             Catalog catalog = catalogService.catalog(catalogVersion);
 
+            // This assert does not make sense, {@link CatalogService#catalog} throws CatalogNotFoundException if the catalog is not found.
             assert catalog != null : "Not found catalog for version " + catalogVersion;
 
-            for (CatalogIndexDescriptor indexDescriptor : catalog.indexes(primaryReplicaId.tableId())) {
+            for (CatalogIndexDescriptor indexDescriptor : catalog.indexes(tableDescriptor.id())) {
                 if (indexDescriptor.status() == BUILDING) {
-                    scheduleBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage, enlistmentConsistencyToken(replicaMeta));
+                    scheduleBuildIndex(
+                            tableDescriptor.zoneId(),
+                            tableDescriptor.id(),
+                            enabledColocation()
+                                    ? ((ZonePartitionId) primaryReplicaId).partitionId()
```

Comment:
In a few places in the code I see this pattern. It's easier to just cast to `PartitionGroupId` and take `partitionId()` from it

Reply:
Nice catch! Thank you!